src/keystore.cpp

```diff
@@ -287,3 +287,108 @@ bool CBasicKeyStore::GetSaplingExtendedSpendingKey(const libzcash::SaplingPaymen
             GetSaplingSpendingKey(extfvk, extskOut);
 }
+
+//
+// Unified Keys
+//
+
+bool CBasicKeyStore::AddUnifiedFullViewingKey(
+        const libzcash::ZcashdUnifiedFullViewingKey &ufvk)
+{
+    LOCK(cs_KeyStore);
+
+    // Add the Sapling component of the UFVK to the wallet.
+    auto saplingKey = ufvk.GetSaplingKey();
+    if (saplingKey.has_value()) {
+        auto ivk = saplingKey.value().fvk.in_viewing_key();
+        mapSaplingKeyUnified.insert(std::make_pair(ivk, ufvk.GetKeyID()));
+    }
+
+    // We can't reasonably add the transparent component here, because
+    // of the way that transparent addresses are generated from the
+    // P2PKH part of the unified address. Instead, whenever a new
+    // unified address is generated, the keys associated with the
+    // transparent part of the address must be added to the keystore.
+
+    // Add the UFVK by key identifier.
+    mapUnifiedFullViewingKeys.insert({ufvk.GetKeyID(), ufvk});
+
+    return true;
+}
+
+bool CBasicKeyStore::AddUnifiedAddress(
+        const libzcash::UFVKId& keyId,
+        const libzcash::diversifier_index_t& diversifierIndex,
+        const libzcash::UnifiedAddress& ua)
+{
+    LOCK(cs_KeyStore);
+
+    // It's only necessary to add p2pkh and p2sh components of
+    // the UA; all other lookups of the associated UFVK will be
+    // made via the protocol-specific viewing key that is used
+    // to trial-decrypt a transaction.
+    auto addrEntry = std::make_pair(keyId, diversifierIndex);
+
+    auto p2pkhReceiver = ua.GetP2PKHReceiver();
+    if (p2pkhReceiver.has_value()) {
+        mapP2PKHUnified.insert(std::make_pair(p2pkhReceiver.value(), addrEntry));
+    }
+
+    auto p2shReceiver = ua.GetP2SHReceiver();
+    if (p2shReceiver.has_value()) {
+        mapP2SHUnified.insert(std::make_pair(p2shReceiver.value(), addrEntry));
+    }
+
+    return true;
+}
+
+std::optional<libzcash::ZcashdUnifiedFullViewingKey> CBasicKeyStore::GetUnifiedFullViewingKey(
+        const libzcash::UFVKId& keyId) const
+{
+    auto mi = mapUnifiedFullViewingKeys.find(keyId);
+    if (mi != mapUnifiedFullViewingKeys.end()) {
+        return mi->second;
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::optional<std::pair<libzcash::UFVKId, std::optional<libzcash::diversifier_index_t>>>
+CBasicKeyStore::GetUFVKMetadataForReceiver(const libzcash::Receiver& receiver) const
+{
+    return std::visit(FindUFVKId(*this), receiver);
+}
+
+std::optional<std::pair<libzcash::UFVKId, std::optional<libzcash::diversifier_index_t>>>
+FindUFVKId::operator()(const libzcash::SaplingPaymentAddress& saplingAddr) const {
+    if (keystore.mapSaplingIncomingViewingKeys.count(saplingAddr) > 0) {
+        const auto& saplingIvk = keystore.mapSaplingIncomingViewingKeys.at(saplingAddr);
+        if (keystore.mapSaplingKeyUnified.count(saplingIvk) > 0) {
+            return std::make_pair(keystore.mapSaplingKeyUnified.at(saplingIvk), std::nullopt);
```

Comment:
Why does this always return `std::nullopt` for the optional diversifier index?

Reply:
I guess that it would be possible for this to decrypt the diversifier to obtain the index and return it. I hadn't done so because the keystore doesn't actually store this information, but maybe it should go ahead and do that, that might simplify things for the caller. Will investigate.

Reply:
Oh, right. It's because the SaplingIncomingViewingKey doesn't actually include `dk`.

Reply:
I guess maybe the right thing to do here would be to walk up to the extfvk and use that to get dk.

Reply:
Okay, it's a bit more complicated, and I'm not sure what the right thing to do is here. The issue is really that the UFVK doesn't have the SaplingExtendedFullViewingKey, so the way by which the SaplingExtendedFullViewingKey gets added to the wallet is separate from the `AddUnifiedFullViewingKey` call - that only gets added when we're adding a unified *spending* key to the wallet.. I didn't modify the keystore to work with SaplingDiversifiableFullViewingKey because I thought that would be too invasive. However, if we want to fully support the import of UFVKs, we'll need to address this issue. I'm inclined to leave this as it is for now, and open a draft PR with the partial fix (which fails tests because it can't actually look up the sapling extfvk to get dk) and figure out how to make that work later.

Reply:
https://github.com/zcash/zcash/pull/5447/commits/a4c316aabf79a585483634f1d8d48e8ff3c5b535 addresses this.